File: contest/ICPC/2018/欧洲/NWERC西北欧/data/jinxedbetting/submissions/accepted/tobi.cpp
#include <iostream>
#include <algorithm>
#include <list>

using namespace std;
typedef long long ll;

ll N, jeff, vals[100000], steps = 0;
list<pair<ll, int>> valFrequency;
list<pair<ll, int>>::iterator getSecond() {
	auto it = valFrequency.begin();
	return ++it;
}

void step(ll todo) {
	auto it = getSecond();
	if (it == valFrequency.end() || it->first + 1 < valFrequency.front().first) {
		valFrequency.insert(it, make_pair(valFrequency.front().first - 1, 0));
		it = getSecond();
	}
	int curLargestCount = valFrequency.front().second;
	ll curLargestVal = valFrequency.front().first;
	if (it->second == 0) {
		int logtwo = 0, tmp = curLargestCount;
		while (tmp >>= 1) logtwo++;
		logtwo++;
		if (++it != valFrequency.end()) todo = min(todo, curLargestVal - it->first);
		if (curLargestCount == 1) todo = curLargestVal - it->first;
		ll howmuch = todo/logtwo - 1;
		if (howmuch > 0) { // combine steps
			steps += logtwo * howmuch;
			jeff -= logtwo * howmuch;
			valFrequency.pop_front();
			valFrequency.front().second += curLargestCount;
			valFrequency.front().first = curLargestVal - howmuch;
			return;
		}
	}
	// single step
	steps++; jeff--;
	if (curLargestCount == 1) {
		valFrequency.pop_front();
		valFrequency.front().second++;
	} else {
		int shift = valFrequency.front().second - valFrequency.front().second / 2;
		valFrequency.front().second /= 2;
		getSecond()->second += shift;
	}
}

int main() {
	cin >> N >> jeff;
	N--;
	for (ll i = 0; i < N; i++) cin >> vals[i];
	sort(vals, vals + N);
	reverse(vals, vals + N);
	ll lastVal = vals[0], count = 0;
	for (ll i = 0; i < N; i++, count++) if (lastVal != vals[i]) {
		valFrequency.push_back(make_pair(lastVal, count));
		lastVal = vals[i], count = 0;
	}
	valFrequency.push_back(make_pair(lastVal, count));
	while (jeff >= valFrequency.front().first) step(jeff - valFrequency.front().first);
	cout << steps-1 << endl;
	return 0;
}

File: contest/ICPC/2018/欧洲/NWERC西北欧/data/jinxedbetting/submissions/dontcare/tobi.java
import java.io.*;
import java.util.*;

public class tobi {
	static int N;
	static long jeff, vals[] = new long[100000], steps = 0;
	static LinkedList<Pair> valFrequency = new LinkedList<>();
	public static void main(String[] args) {
		Scanner scanner = new Scanner(System.in);
		N = scanner.nextInt() - 1;
		jeff = scanner.nextLong();
		for (int i = 0; i < N; i++) vals[i] = scanner.nextLong();
		Arrays.sort(vals, 0, N);
		for (int i = 0; i < N/2; i++) {
			long tmp = vals[i];
			vals[i] = vals[N-1-i];
			vals[N-1-i] = tmp;
		}
		long lastVal = vals[0];
		int count = 0;
		for (int i = 0; i < N; i++, count++) if (lastVal != vals[i]) {
			valFrequency.add(new Pair(lastVal, count));
			lastVal = vals[i]; count = 0;
		}
		valFrequency.add(new Pair(lastVal, count));
		while (jeff >= valFrequency.peek().first) step(jeff - valFrequency.peek().first);
		System.out.println(steps-1);
	}
	static void step(long todo) {
		if (valFrequency.size() == 1 || valFrequency.get(1).first + 1 < valFrequency.peek().first) {
			valFrequency.add(1, new Pair(valFrequency.peek().first - 1, 0));
		}
		int curLargestCount = valFrequency.peek().second;
		long curLargestVal = valFrequency.peek().first;
		if (valFrequency.get(1).second == 0) {
			int logtwo = 0, tmp = curLargestCount;
			while (tmp >> 1 > 0) {
				logtwo++;
				tmp = tmp >> 1;
			}
			logtwo++;
			if (valFrequency.size() > 2) {
				todo = Math.min(todo, curLargestVal - valFrequency.get(2).first);
				if (curLargestCount == 1) todo = curLargestVal - valFrequency.get(2).first;
			}
			long howmuch = todo/logtwo - 1;
			if (howmuch > 0) { // combine steps
				steps += logtwo * howmuch;
				jeff -= logtwo * howmuch;
				valFrequency.pop();
				valFrequency.peek().second += curLargestCount;
				valFrequency.peek().first = curLargestVal - howmuch;
				return;
			}
		}
		// single step
		steps++; jeff--;
		if (curLargestCount == 1) {
			valFrequency.pop();
			valFrequency.peek().second++;
		} else {
			int shift = valFrequency.peek().second - valFrequency.peek().second / 2;
			valFrequency.peek().second /= 2;
			valFrequency.get(1).second += shift;
		}
	}
}

class Pair {
	public Long first;
	public Integer second;
	public Pair(Long first, Integer second) {
		this.first = first;
		this.second = second;
	}
}
